### PR TITLE
feat(dotnet-sdk): Use On-Behalf-Of flow of MSAL

### PR DESCRIPTION
**Pull request by tecton**
- [x] 1. Remove the runtime dependency of simple auth service.
- [x] 2. Use MSAL for On-Behalf-Of flow to exchange access token.
- [x] 3. Bump version to 0.4.0-rc.
- [x] 4. Update README and upgrade steps.

Breaking change:
- Add required configuration field `ClientSecret` and `OAuthAuthority`.

E2E TEST: not available.

### CODE REVIEW

File: packages/dotnet-sdk/src/TeamsFx/Credential/TeamsUserCredential.cs [hunk 1]
```diff
@@ -344,186 +353,59 @@ private async Task<AccessToken> GetSsoTokenAsync()
...
-            new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json")).ConfigureAwait(false);
-        var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
-        if (response.IsSuccessStatusCode)
+
+        try
         {
-            var oauthToken = JsonSerializer.Deserialize<SimpleAuthAccessToken>(content, new JsonSerializerOptions { IncludeFields = true });
-            var accessToken = new AccessToken(oauthToken.access_token, oauthToken.expires_on);
+            _logger.LogDebug("Acquiring token via OBO flow.");
+            var result = await _identityClientAdapter
```
[the remark this answers is not in the document]
Comment (tecton) [reply packages/dotnet-sdk/src/TeamsFx/Credential/TeamsUserCredential.cs:368]: Thanks, if the access token is cached in MSAL, we can remove the memory cache.

File: packages/dotnet-sdk/README.md [hunk 2]
```diff
@@ -38,8 +38,9 @@ For .NET 6 projects (VS 2022): Choose version >= 0.3.0-rc.
 "TeamsFx": {
     "Authentication": {
         "ClientId": "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx",
-        "SimpleAuthEndpoint": "https://localhost:44357/",
-        "InitiateLoginEndpoint": "https://localhost:44357/auth-start.html"
+        "ClientSecret": "xxx", // 'User Secrets' is a better place to store secret string.
+        "InitiateLoginEndpoint": "https://localhost:44357/auth-start.html",
```
Comment (JerryYangKai) [inline packages/dotnet-sdk/README.md:42]: Our future default port is 44302, avoid 44357 to cause some unexpected problem.

File: packages/dotnet-sdk/README.md [hunk 3]
```diff
@@ -135,6 +137,17 @@ private void printLog(LogLevel level, string message)
 }
 ```
 
+### Upgrade from 0.3.0-rc to 0.4.0 (For projects created by Visual Studio 2022 17.1 Preview toolkit)
+If there is an existing project created in VS2022 17.1 Preview, you can use the following steps to upgrade:
+- In `appsettings.{Environment}.json` file:
+1. Add `OAuthAuthority` under `TeamsFx:Authentication` and copy the value from `OAUTH_AUTHORITY`.
+2. Remove the line `"SimpleAuthEndpoint": "https://localhost:44357/"`.
```
Comment (JerryYangKai) [inline packages/dotnet-sdk/README.md:144]: Maybe this line could be 
Remove the line `"SimpleAuthEndpoint": "https://localhost:port/"`.

**Review by JerryYangKai — DISMISSED** [review]

File: packages/dotnet-sdk/README.md (hunk 2 above)
Comment (tecton) [reply packages/dotnet-sdk/README.md]: Thanks for the reminder.

**Review by JerryYangKai — APPROVED** [review]

**Pull request merged**